### ISSUE TITLE
Programmatic API access to information printed by Pkg.status()

### ISSUE THREAD

**Issue by mbauman** [JuliaLang/Pkg.jl#2238]
It'd be really nice to actually return a value from `Pkg.status` that is then displayed (instead of having it print as a side-effect).  The motivating use-case here for me is to request the currently installed version of a package.  This has been a [question on stack overflow](https://stackoverflow.com/questions/25635508/determine-version-of-a-specific-package) for a long time and doesn't have a good answer.

* `Pkg.installed()` is deprecated
* `Pkg.status("Example")` prints and isn't programmatically accessible
* @fredrikekre suggests on Slack to use `Pkg.dependencies()[Base.UUID("7876af07-990d-54b4-ab0e-23690620f79a")].version`, but that additionally requires an API to convert a module or package name to its UUID.

**Comment by KristofferC** [comment]
> But that additionally requires an API to convert a module or package name to its UUID.

So you only want to query something that is a direct dependency? Otherwise, the UUID is needed to disambiguate potentially duplicated names. In that case, looking at

https://julialang.github.io/Pkg.jl/dev/api/#Pkg.dependencies

you can look for the pkg with `is_direct_dep == true` and with the correct name.

**Comment by mbauman** [comment]
Yes, limiting this to direct dependencies would be sensible for my use-case.

The request here is for a higher level API as simple as `Pkg.status("Example")`.

**Comment by KristofferC** [comment]
Oh, I thought you just wanted to know how to do it.

I guess a one-line wrapper around `Pkg.dependencies` could be added to Pkg, although it feels maybe a bit excessive.

**Comment by oscardssmith** [comment]
Can we please do this? It would be so much nicer if `Pkg.status(x)` returned a `PackageSpec`

**Comment by KristofferC** [comment]
As been said, this already exists. A PackageSpec is not really suits le because that an it eg hold a git tree sha. Is what is needed a way to lookup a uuid from a name in the project file?

**Comment by mhauru** [comment]
> Is what is needed a way to lookup a uuid from a name in the project file?

In my mind the request is simpler: For `Pkg.status` to return a value that contains the same information as what is printed out. It just seems silly to not have programmatic access to the information that `Pkg.status` prints out without any extra trickery other than `x = Pkg.status("MyFavouritePackage")`.

Commenting because I recently yearned for this feature, for reasons I've now forgotten.

**Comment by KristofferC** [comment]
>  For Pkg.status to return a value that contains the same information as what is printed out.

Isn't that mostly `Pkg.dependencies()`? What more information do you want?

**Comment by oscardssmith** [comment]
The main issue with `dependencies()` is that it's keyed by UUID rather than name which makes it really annoying if you want to figure out what version a package is.

**Comment by KristofferC** [comment]
You can just filter it based on name then? Package names are not unique in an environment so it kind of has to be like that.

**Comment by oscardssmith** [comment]
What you're saying is that the ideal interface to look up the version of LinearAlgebra is `only(p for p in values(Pkg.dependencies()) if p.name == "LinearAlgebra").version` rather than `Pkg.status("LinearAlgebra").version`? Like the `Pkg.dependencies()` approach works, but it's O(n), and seems overly cumbersome when `Pkg.status` could just return the same thing.

**Comment by KristofferC** [comment]
Maybe you missed 

> Package names are not unique in an environment so it kind of has to be like that.

If you want specifically the stdlib LinearAlgebra you would have to use the UUID to be sure you get the right package (that's how that package is uniquely identified).

Anyway, I don't think you should have to depend on a package manager to do this very basic environment queries. A separate small package for that seems much better.

**Comment by oscardssmith** [comment]
I understand that this theoretically is an issue, but `Pkg.add`, `Pkg.status` and most other user facing functions allow the use of names rather than UUIDs (and throw an error if this is ambiguous).

> Anyway, I don't think you should have to depend on a package manager to do these very basic environment queries. A separate small package for that seems much better.

Totally agreed. It would be great if we had a stdlib that exposed `Pkg.status`  (and possibly other Pkg functionality that doesn't require downloading Pkgs) that was a lot lighter that could be relied upon.

**Comment by mhauru** [comment]
> Isn't that mostly Pkg.dependencies()? What more information do you want?

I want the same information, but from the function that I would expect to get it from. It's not about whether the information is available, it's about having a good user interface. To run `Pkg.status()`, see the information I need in the REPL, but to have no way of programmatically accessing that information without pulling out the manual to figure out what _other_ function would give me what I need is a bad user experience. The first time I found myself in that situation I did not even know `Pkg.dependencies` existed. It's unintuitive and surprising that `Pkg.status()` doesn't give you a return value that would have the same information that it prints out, especially since it doesn't use its return value for anything else either.

**Comment by KristofferC** [comment]
>  It's unintuitive and surprising that Pkg.status() doesn't give you a return value 

You need to return almost everything Pkg knows about to be able to replicate what `Pkg.status` shows. It shows things like outdated packages (that requires access to the registry), it can show a diff towards the latest commit (so requires access to the git repo of the current project) etc etc. There is no useful return value to return and just returning a mega object with everything isn't how you write good APIs. 

`Pkg.dependencies` gives information about packages in the current environment, I think that is exactly what is desired here.

For registry information there is RegistryInstances.jl since we haven't wanted to commit ourselves to a public API for registry information.

**Comment by mhauru** [comment]
> There is no useful return value to return and just returning a mega object with everything isn't how you write good APIs.

The useful return value would be a simple data structure that provides programmatic access to the data printed. The print-out is essentially either a table, or with `extensions=true` a depth-two tree. These would be very easy to turn into something like a list or a dictionary of (Named)Tuples, or a purpose-specific struct that does the same, or whatever representation of simple tables/trees you prefer.

@KristofferC do you not agree that it's a bad user experience to have a command available to print out some piece of useful information to the REPL, but not be able to use the same command to get the same information as a value? This seems obvious to me, and thus I find the resistance hard to understand.

Commenting because I again today did `] status`, saw the information I needed, and then had to break my flow and go googling to remind myself how to get that information into a variable.

**Comment by KristofferC** [comment]
> The useful return value would be a simple data structure that provides programmatic access to the data printed.

Ok, this data structure is `Pkg.Types.Context()`.

**Comment by DilumAluthge** [comment]
> saw the information I needed

I think it would be helpful for us to know what the specific pieces of information were that you needed for your use case. As Kristoffer mentioned, `Pkg.dependencies()` can already provide information on what packages you have, and RegistryInstances.jl can provide registry information.

**Comment by KristofferC** [comment]
> do you not agree that it's a bad user experience to have a command available to print out some piece of useful information to the REPL, but not be able to use the same command to get the same information as a value?

No, I do not, and I've already stated my reasons for doing so. To reiterate, `Pkg.status` needs access to basically every single piece of information that Pkg has access to. Bundling that up into some mega object and giving it to users is a badly designed API. The API for users should be more specific, for example, one set of methods that handles registries, one set that handles environments, etc. And then users can create whatever higher-level functionality they want on top of that. Returning some object for every function that happens to print something is not a good idea. 

If you want some analogous you can look at `Profile.print` or `code_llvm`.

**Comment by oscardssmith** [comment]
> Returning some object for every function that happens to print something is not a good idea.

I strongly disagree with this. Sure, for more efficient coding you want specific APIs, but I can see no reason that the REPL APIs shouldn't return data structures with printing methods to display the output.

**Comment by IanButterworth** [comment]
That's more Julian, yeah, but the type it returns doesn't need to be public for that to work. This is about having a public API.

**Comment by KristofferC** [comment]
> but I can see no reason that the REPL APIs shouldn't return data structures with printing methods to display the output.

🤦  I've already stated the reason multiple times. The object is `Pkg.Types.Context`. Here you can see it is the input to the status function: https://github.com/JuliaLang/Pkg.jl/blob/e749a514020e45d71362ba55cf86e764feaece18/src/API.jl#L1373. So use that. Are you happy with that answer? Probably not (which is exactly the point I am trying to make).

To try to make the point even clearer. There should be a set of orthogonal API methods that can provide the information people might want to retrieve (about registries, environments, etc). Pkg itself, for example, uses that API to create a `status` function which prints various information about the environment to the terminal. But `status` itself is not the API! The output of that is intended for human (or perhaps LLM) consumption. It can be changed between releases, it can be made to look different etc. Tying the `status` output to some object that is interacted with programmatically is mixing up responsibilities.

**Comment by mhauru** [comment]
I feel like we've been talking past each other a bit, with both sides feeling like the other side is ignoring their arguments and just repeating the same points. This is starting to make a bit more sense to me now.

Mine (and I think Oscar's) point is that if I can get the information delivered to my REPL, it feels silly and frustrating to not have it available as a value. "Clearly there is a table-like value `x` and I'm looking at `print(x)`, but just give me `x`!" The extreme here would be for me to write a regexp to parse the string that `Pkg.status` prints, but obviously no one thinks that's a good idea. Yes, the information is available through other means (e.g. `Pkg.dependencies`), but if I already have one way of getting the information I need I don't want to go and find another one (most users don't know most of the `Pkg` API), and smooth user experiences matter.

Your point seems to be that if we make `status` return anything then that's part of the public API and you'd feel obligated to keep it stable. (I'm less convinced by the argument that the return type needs to be super complex, because in my mind the goal is to just return a value with exactly the same information that is printed, nothing more, and what is printed is not super complex.) That's a fair point, and I'm not quite sure what to do about it. For the use cases I have, I would be fine with the `Pkg.status` return value changing as the print out changes and expect the two to mirror each other 1-to-1 (in a way that I would not be fine with `Pkg.dependencies` changing), but I don't know if there's a good way to communicate that and leave it outside of the stable API.

**Comment by KristofferC** [comment]
> because in my mind the goal is to just return a value with exactly the same information that is printed, nothing more, and what is printed is not super complex

I mean, we can play the game:

```
(@v1.12) pkg> status
Status `~/.julia/environments/v1.12/Project.toml`
  [6e4b80f9] BenchmarkTools v1.6.0
  [f68482b8] Cthulhu v2.17.6
```

Ok, so maybe I would have:

```
struct Package
    uuid::UUID
    name::String
    version::VersionNumber
    loc::String # url or path if it is e.g. devved
end

struct StatusObject
    active_project::String
    packages::Vector{Package}
end
```

but then I happen to see the

```
⌃ [295af30f] Revise v3.8.0
```

line. So I need to add a `upgradable::Bool` to `Package`? But then I can write

```
(@v1.12) pkg> st --outdated
Status `~/.julia/environments/v1.12/Project.toml`
⌃ [31a5f54b] Debugger v0.7.13 (<v0.7.15)
⌃ [295af30f] Revise v3.8.0 (<v3.9.0)
```

So I need to add a `max_version` to `Package`. But, `st --outdated` can also show what packages are holding something back, so it should have a `Vector{Package}` of packages holding it back? But then we can do `st --diff` to show the diff vs the previous commit. So now I need a `prev_packages::Vector{Package}` in `StatusObject`? Oh, and I also need to know if the package is a direct dependency or not. Oh, and we also have `--workspace` so we need to probably put a `workspace::Vector{Project}` in the `StatusObject`. And there is `-e` to show all the extension information which has to go in there somewhere. It becomes unwieldly and inefficient to have to shove all of this into one thing.
  
To give you a web analogy, the output of `Pkg.status` is a "front end" thing. It is akin to a rendered website (HTML). You can look at it and use it directly, but if you want to actually start interacting with the data of the site programmatically, you would use e.g. a REST API for it. You don't want how the website looks to influence your programmatic interaction with it because some designer might come and redesign the whole website. Just like someone might come and redesign `Pkg.status` to look completely different. But you still want your code to keep working. That's why you use a different abstraction layer (like a REST API) for the programmatic interactions. In other words, the rendering of `Pkg.status` is not something that should be relied upon for writing code. Returning something that has a "one-to-one" correspondence to what is printed would give the impression that that is a good idea.

If there are specific queries that you want to do that are currently not possible, bring those up, and they can possibly be added.
